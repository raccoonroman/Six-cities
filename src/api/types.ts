import { OfferCategory, Location } from '@/types';

export interface AuthInfoRaw {
  avatar_url: string;
  email: string;
  id: number;
  is_pro: boolean;
  name: string;
}

export interface User {
  email: string;
  password: string;
}

export interface OfferRaw {
  id: number;
  title: string;
  preview_image: string;
  price: number;
  rating: number;
  type: OfferCategory;
  bedrooms: number;
  max_adults: number;
  is_favorite: boolean;
  is_premium: boolean;
  location: Location;
  city: {
    name: string;
    location: Location;
  };
  description: string;
  goods: string[];
  host: {
    avatar_url: string;
    id: number;
    is_pro: boolean;
    name: string;
  };
  images: string[];
}

export interface CommentRaw {
  id: number;
  comment: string;
  date: string;
  rating: number;
  user: {
    avatar_url: string;
    id: number;
    is_pro: boolean;
    name: string;
  };
}

export interface CommentPost {
  comment: string;
  rating: number;
}
