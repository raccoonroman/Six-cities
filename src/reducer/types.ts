import { OfferRaw } from '@/api/types';
import { ActionType } from '@/const';

interface LoadOffersAction {
  type: typeof ActionType.LOAD_OFFERS
  payload: OfferRaw[]
}

interface UpdateOfferAction {
  type: typeof ActionType.UPDATE_OFFER
  payload: OfferRaw
}

interface SetEmailAction {
  type: typeof ActionType.SET_EMAIL
  payload: string
}

export type offersAction = LoadOffersAction | UpdateOfferAction;
export type userDataAction = SetEmailAction;
