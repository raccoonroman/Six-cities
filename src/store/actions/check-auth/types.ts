enum CheckAuth {
  PENDING = 'CHECK_AUTH/PENDING',
  RESOLVE = 'CHECK_AUTH/RESOLVE',
  REJECT = 'CHECK_AUTH/REJECT',
}

export default CheckAuth;
