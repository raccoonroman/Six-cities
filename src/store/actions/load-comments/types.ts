enum LoadComments {
  PENDING = 'LOAD_COMMENTS/PENDING',
  RESOLVE = 'LOAD_COMMENTS/RESOLVE',
  REJECT = 'LOAD_COMMENTS/REJECT',
}

export default LoadComments;
