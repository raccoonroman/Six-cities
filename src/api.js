import axios from 'axios';


const Error = {
  UNAUTHORIZED: 401,
};

export const createAPI = (onUnauthorized) => {
  const api = axios.create({
    baseURL: `https://htmlacademy-react-3.appspot.com/six-cities`,
    timeout: 5000,
    withCredentials: true,
  });

  const onSuccess = ({data}) => data;

  const onFail = (err) => {
    const {status} = err.response;

    if (status === Error.UNAUTHORIZED) {
      onUnauthorized();
      throw err;
    }

    throw err;
  };

  api.interceptors.response.use(onSuccess, onFail);

  return api;
};
