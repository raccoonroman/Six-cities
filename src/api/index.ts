import HttpClient from '@/api/http-client';
import {
  AuthInfoRaw,
  User,
  OfferRaw,
  CommentRaw,
  CommentPost,
} from '@/api/types';

export default class Api extends HttpClient {
  public constructor(onUnauthorized: Function) {
    super('https://htmlacademy-react-3.appspot.com/six-cities', onUnauthorized);
  }

  public checkAuth = () => this.instance.get<AuthInfoRaw>('/login');

  public login = (loginData: User) => (
    this.instance.post<AuthInfoRaw>('/login', loginData));

  public loadOffers = () => this.instance.get<OfferRaw[]>('/hotels');

  public loadNearbyOffers = (offerId: number) => (
    this.instance.get<OfferRaw[]>(`/hotels/${offerId}/nearby`));

  public loadComments = (offerId: number) => (
    this.instance.get<CommentRaw[]>(`/comments/${offerId}`));

  public postComment = (commentData: CommentPost, offerId: number) => (
    this.instance.post<CommentRaw[]>(`/comments/${offerId}`, commentData));

  public setFavoriteStatus = (offerId: number, status: number) => (
    this.instance.post<OfferRaw>(`/favorite/${offerId}/${status}`));
}
