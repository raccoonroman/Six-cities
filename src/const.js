export const DEFAULT_CITIES = [`Paris`, `Cologne`, `Brussels`, `Amsterdam`, `Hamburg`, `Dusseldorf`];

export const SortType = {
  POPULAR: `Popular`,
  PRICE_LOW_TO_HIGH: `Price: low to high`,
  PRICE_HIGH_TO_LOW: `Price: high to low`,
  TOP_RATED_FIRST: `Top rated first`,
};

export const CardType = {
  CITY: `city`,
  NEAR: `near`,
  FAVORITE: `favorite`,
};

export const ActionType = {
  LOAD_OFFERS: `LOAD_OFFERS`,
  UPDATE_OFFER: `UPDATE_OFFER`,
  SET_CITY: `SET_CITY`,
  REQUIRED_AUTHORIZATION: `REQUIRED_AUTHORIZATION`,
  SET_EMAIL: `SET_EMAIL`,
  LOAD_COMMENTS: `LOAD_COMMENTS`,
  LOAD_NEARBY_OFFERS: `LOAD_NEARBY_OFFERS`,
};

export const AuthorizationStatus = {
  AUTH: `AUTH`,
  NO_AUTH: `NO_AUTH`,
};

export const AppRoute = {
  ROOT: `/`,
  LOGIN: `/login`,
  FAVORITES: `/favorites`,
  OFFER: `/offer`,
};
