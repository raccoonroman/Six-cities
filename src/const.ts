export const DEFAULT_CITIES: string[] = ['Paris', 'Cologne', 'Brussels', 'Amsterdam', 'Hamburg', 'Dusseldorf'];

export enum SortType {
  POPULAR = 'Popular',
  PRICE_LOW_TO_HIGH = 'Price: low to high',
  PRICE_HIGH_TO_LOW = 'Price: high to low',
  TOP_RATED_FIRST = 'Top rated first',
}

export enum CardType {
  CITY = 'city',
  NEAR = 'near',
  FAVORITE = 'favorite',
}

export enum ActionType {
  LOAD_OFFERS = 'LOAD_OFFERS',
  UPDATE_OFFER = 'UPDATE_OFFER',
  SET_CITY = 'SET_CITY',
  REQUIRED_AUTHORIZATION = 'REQUIRED_AUTHORIZATION',
  SET_EMAIL = 'SET_EMAIL',
  LOAD_COMMENTS = 'LOAD_COMMENTS',
  LOAD_NEARBY_OFFERS = 'LOAD_NEARBY_OFFERS',
}

export enum AuthorizationStatus {
  AUTH = 'AUTH',
  NO_AUTH = 'NO_AUTH',
}

export enum AppRoute {
  ROOT = '/',
  LOGIN = '/login',
  FAVORITES = '/favorites',
  OFFER = '/offer',
}

export enum MapType {
  HOVERED_ACTIVE_OFFER = 'HOVERED_ACTIVE_OFFER',
  STATIC_ACTIVE_OFFER = 'STATIC_ACTIVE_OFFER',
}
