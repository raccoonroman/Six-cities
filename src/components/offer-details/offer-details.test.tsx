import React from 'react';
import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import { MemoryRouter, Route } from 'react-router-dom';
import { AuthorizationStatus, AppRoute } from '@/const';
import Api from '@/api';
import { OfferRaw } from '@/api/types';
import OfferDetails from '@/components/offer-details';


const onUnauthorized = () => {
  // do nothing
};

const api = new Api(onUnauthorized);
const middlewares = [thunk.withExtraArgument(api)];
const mockStore = configureStore(middlewares);

configure({ adapter: new Adapter() });


const mockOffers: OfferRaw[] = [
  {
    id: 100490,
    title: 'The best title ever',
    preview_image: 'https://i.picsum.photos/id/22/400/200.jpg',
    price: 333,
    rating: 3.3,
    type: 'apartment',
    bedrooms: 12,
    max_adults: 12,
    is_favorite: false,
    is_premium: false,
    location: {
      latitude: 52.35,
      longitude: 4.67,
      zoom: 14,
    },
    city: {
      name: 'Vinnytsya',
      location: {
        latitude: 52.370216,
        longitude: 4.895168,
        zoom: 12,
      },
    },
    description: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.',
    goods: ['Baby seat', 'Towels', 'Dishwasher', 'Breakfast'],
    host: {
      avatar_url: 'https://i.picsum.photos/id/58/400/200.jpg',
      id: 911,
      is_pro: false,
      name: 'Rachel',
    },
    images: ['https://i.picsum.photos/id/82/400/200.jpg', 'https://i.picsum.photos/id/83/400/200.jpg', 'https://i.picsum.photos/id/84/400/200.jpg', 'https://i.picsum.photos/id/85/400/200.jpg', 'https://i.picsum.photos/id/86/400/200.jpg', 'https://i.picsum.photos/id/87/400/200.jpg', 'https://i.picsum.photos/id/88/400/200.jpg', 'https://i.picsum.photos/id/89/400/200.jpg'],
  },
  {
    id: 100491,
    title: 'Some title',
    preview_image: 'https://i.picsum.photos/id/23/400/200.jpg',
    price: 900,
    rating: 4.5,
    type: 'room',
    bedrooms: 2,
    max_adults: 4,
    is_favorite: true,
    is_premium: true,
    location: {
      latitude: 51,
      longitude: 5,
      zoom: 16,
    },
    city: {
      name: 'Vinnytsya',
      location: {
        latitude: 52,
        longitude: 4.9,
        zoom: 12,
      },
    },
    description: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. ',
    goods: ['Towels', 'Breakfast'],
    host: {
      avatar_url: 'https://i.picsum.photos/id/59/400/200.jpg',
      id: 912,
      is_pro: true,
      name: 'Chandler',
    },
    images: ['https://i.picsum.photos/id/182/400/200.jpg', 'https://i.picsum.photos/id/183/400/200.jpg', 'https://i.picsum.photos/id/184/400/200.jpg', 'https://i.picsum.photos/id/185/400/200.jpg', 'https://i.picsum.photos/id/186/400/200.jpg', 'https://i.picsum.photos/id/187/400/200.jpg', 'https://i.picsum.photos/id/188/400/200.jpg', 'https://i.picsum.photos/id/189/400/200.jpg'],
  },
];


it('Render <OfferDetails />', () => {
  const store = mockStore({
    offers: {
      loadOffersStatus: {
        pending: false,
        resolve: false,
        reject: false,
      },
      offers: mockOffers,
    },
    authorization: {
      authorization: AuthorizationStatus.AUTH,
    },
    userData: {
      email: 'name@gmail.com',
    },
    comments: {
      postCommentStatus: {
        pending: false,
        resolve: false,
        reject: false,
      },
      comments: [],
    },
    nearbyOffers: {
      nearbyOffers: mockOffers,
    },
  });

  const tree = mount(
    <Provider store={store}>
      <MemoryRouter initialEntries={[`${AppRoute.OFFER}/100491`]}>
        <Route
          exact
          path={`${AppRoute.OFFER}/:id`}
          component={OfferDetails}
        />
      </MemoryRouter>
    </Provider>,
  );

  expect(tree.getDOMNode()).toMatchSnapshot();
});
