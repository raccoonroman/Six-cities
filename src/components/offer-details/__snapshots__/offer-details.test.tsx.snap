// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render <OfferDetails /> 1`] = `null`;
